spring-cloud-aws-cloudmap/src/main/java/org/springframework/cloud/aws/cloudmap/CloudMapRegistryService.java

```diff
@@ -0,0 +1,355 @@
+import com.amazonaws.services.servicediscovery.model.ListNamespacesRequest;
+import com.amazonaws.services.servicediscovery.model.ListServicesRequest;
+import com.amazonaws.services.servicediscovery.model.NamespaceAlreadyExistsException;
+import com.amazonaws.services.servicediscovery.model.NamespaceFilter;
+import com.amazonaws.services.servicediscovery.model.NamespaceSummary;
+import com.amazonaws.services.servicediscovery.model.Operation;
+import com.amazonaws.services.servicediscovery.model.RecordType;
+import com.amazonaws.services.servicediscovery.model.RegisterInstanceRequest;
+import com.amazonaws.services.servicediscovery.model.ResourceLimitExceededException;
+import com.amazonaws.services.servicediscovery.model.ServiceAlreadyExistsException;
+import com.amazonaws.services.servicediscovery.model.ServiceFilter;
+import com.amazonaws.services.servicediscovery.model.ServiceSummary;
+import com.amazonaws.util.StringUtils;
+import org.apache.catalina.connector.Connector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.boot.web.embedded.tomcat.TomcatConnectorCustomizer;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.event.ContextClosedEvent;
+
+/**
+ * Contains all the methods to registry with cloudmap.
+ *
+ * @author Hari Ohm Prasath
+ * @since 1.0
+ */
+public class CloudMapRegistryService implements TomcatConnectorCustomizer, ApplicationListener<ContextClosedEvent> {
+
+	private static final String SUBMITTED = "SUBMITTED";
+
+	private static final String PENDING = "PENDING";
+
+	private static final int MAX_POLL = 30;
+
+	private static final String AWS_INSTANCE_IPV_4 = "AWS_INSTANCE_IPV4";
+
+	private static final String REGION = "REGION";
+
+	private static final Logger log = LoggerFactory.getLogger(CloudMapRegistryService.class);
+
+	private final AWSServiceDiscovery serviceDiscovery;
+
+	private final CloudMapRegistryProperties properties;
+
+	private String serviceInstanceId;
+
+	private String serviceId;
+
+	private final CloudMapUtils cloudMapUtils = new CloudMapUtils();
+
+	private volatile Connector connector;
+
+	public CloudMapRegistryService(AWSServiceDiscovery serviceDiscovery, CloudMapRegistryProperties properties) {
+		this.serviceDiscovery = serviceDiscovery;
+		this.properties = properties;
+	}
+
+	/**
+	 * Register with cloudmap, the method takes care of the following: 1. Create
+	 * namespace, if not exists 2. Create service, if not exists 3. Register the instance
+	 * with the created namespace and service
+	 * @return cloudmap registration operation ID
+	 */
+	public String registerInstances() {
+		if (properties != null && !StringUtils.isNullOrEmpty(properties.getNameSpace())
+				&& !StringUtils.isNullOrEmpty(properties.getService())) {
+
+			final String nameSpace = properties.getNameSpace();
+			final String service = properties.getService();
+
+			this.serviceInstanceId = UUID.randomUUID().toString();
```

Comment:
Is it the case that there is always a single instance per service? I am surprised that `registerInstances` actually registers a single instance only.

Reply:
Let's say we have two tasks running per service, this piece of code will create one Cloudmap service and registers two HTTP instances inside the service. There can be only one service and multiple HTTP instances gets registered inside that.

Reply:
I see. I wondered why there is `registerInstances` (plural) and `deregisterInstance` `singular`. That sounds like instances may be left behind. But now I see it's actually just a naming issue.

Reply:
To avoid further confusion, I have renamed the method to `registerInstance()`